shared-module/gamepad/__init__.c

```diff
@@ -27,32 +27,37 @@
 #include <stdbool.h>
 
 #include "py/mpstate.h"
-#include "__init__.h"
-#include "GamePad.h"
+#include "shared-bindings/gamepad/__init__.h"
+#include "shared-bindings/gamepad/GamePad.h"
 
 #include "shared-bindings/digitalio/DigitalInOut.h"
 
 
 void gamepad_tick(void) {
-    gamepad_obj_t* gamepad_singleton = MP_STATE_VM(gamepad_singleton);
-    if (!gamepad_singleton) {
+    uint8_t current = 0;
+    uint8_t bit = 1;
+
+    void* singleton = MP_STATE_VM(gamepad_singleton);
+    if (!singleton) {
         return;
     }
-    uint8_t gamepad_current = 0;
-    uint8_t bit = 1;
-    for (int i = 0; i < 8; ++i) {
-        digitalio_digitalinout_obj_t* pin = gamepad_singleton->pins[i];
-        if (!pin) {
-            break;
-        }
-        if (common_hal_digitalio_digitalinout_get_value(pin)) {
-            gamepad_current |= bit;
+    if (MP_OBJ_IS_TYPE(MP_OBJ_FROM_PTR(singleton), &gamepad_type)) {
+        // buttons connected directly to pins
+        gamepad_obj_t *self = singleton;
+        for (int i = 0; i < 8; ++i) {
+            digitalio_digitalinout_obj_t* pin = self->pins[i];
+            if (!pin) {
+                break;
+            }
+            if (common_hal_digitalio_digitalinout_get_value(pin)) {
+                current |= bit;
+            }
+            bit <<= 1;
         }
-        bit <<= 1;
+        current ^= self->pulls;
+        self->pressed |= self->last & current;
+        self->last = current;
```

Comment:
I wonder why you brought back last as part of the gamepad object, instead of keeping is a static variable here?

Reply:
I really dislike static state within a function because it is not object oriented and not obviously global state. Having it on the object is the common style within CircuitPython.

Reply:
Fair enough.